phabricatoremails/render.py

```diff
@@ -0,0 +1,420 @@
+    SecureRevisionUpdated,
+    SecureRevisionAbandoned,
+    SecureRevisionReclaimed,
+    SecureRevisionCreated,
+)
+from phabricatoremails.template import TemplateStore
+from sqlalchemy.orm import Session
+from statsd import StatsClient
+
+
+class ParseError(Exception):
+    pass
+
+
+def parse_recipient(recipient):
+    if not recipient:
+        return None
+
+    seconds = recipient["timezoneOffset"]  # e.g.: -25200
+    if seconds > 0:
+        hours = floor(seconds / 3600)
+        minutes = floor((seconds - hours * 3600) / 60)
+    else:
+        hours = ceil(seconds / 3600)
+        minutes = ceil((seconds - hours * 3600) / 60)
+    return Recipient(
+        recipient["email"],
+        timezone(timedelta(hours=hours, minutes=minutes)),
+        recipient["isActor"],
+    )
+
+
+def parse_revision(raw_revision):
+    raw_bug = raw_revision.get("bug")
+    bug = Bug(raw_bug["bugId"], raw_bug["name"], raw_bug["link"]) if raw_bug else None
+    return Revision(
+        raw_revision["revisionId"], raw_revision["name"], raw_revision["link"], bug
+    )
+
+
+def parse_secure_revision(raw_revision):
+    raw_bug = raw_revision.get("bug")
+    bug = SecureBug(raw_bug["bugId"], raw_bug["link"]) if raw_bug else None
+    return SecureRevision(raw_revision["revisionId"], raw_revision["link"], bug)
+
+
+def parse_comment_context(context_kind, context):
+    if context_kind == "code":
+        return CodeContext(
+            [
+                DiffLine(
+                    line["lineNumber"], DiffLineType(line["type"]), line["rawContent"]
+                )
+                for line in context["diff"]
+            ]
+        )
+    elif context_kind == "reply":
+        return ReplyContext(
+            context["otherAuthor"],
+            datetime.fromisoformat(context["otherDateUtc"]),
+            context["otherComment"],
+        )
+    else:
+        raise ParseError("Comment context was not code or a reply")
+
+
+def parse_inline_comment(raw_inline):
+    return InlineComment(
+        raw_inline["fileContext"],
+        raw_inline["link"],
+        raw_inline["text"],
+        parse_comment_context(raw_inline["contextKind"], raw_inline["context"]),
+    )
+
+
+def parse_affected_file(raw_file):
+    return AffectedFile(raw_file["path"], AffectedFileChange(raw_file["change"]))
+
+
+def parse_reviewer(raw_reviewer):
+    return Reviewer(
+        raw_reviewer["name"],
+        raw_reviewer["isActionable"],
+        ReviewerStatus(raw_reviewer["status"]),
+        parse_recipient(raw_reviewer.get("recipient")),
+    )
+
+
+def parse_metadata_edited_reviewer(raw_reviewer):
+    return MetadataEditedReviewer(
+        raw_reviewer["name"],
+        raw_reviewer["isActionable"],
+        ReviewerStatus(raw_reviewer["status"]),
+        ExistenceChange(raw_reviewer["metadataChange"]),
+        parse_recipient(raw_reviewer.get("recipient")),
+    )
+
+
+def render(logger: Logger, stats: StatsClient, events: List, db: Session, css_inline):
```

Comment:
Yeah, I'm not happy with the SQLAlchemy naming here. I can call it `session`, but then the name is ambiguous when used down in the function body: "is it an HTTP session? A database session? An Amazon session?".

I also considered `db_session`, but that seemed too heavyweight.
I see what you mean (especially since there's a `Db` class), but I found that `db` struck the right balance between ambiguity and concise-ness.
Would you rather `db_session`?

Reply:
> Yeah, I'm not happy with the SQLAlchemy naming here. I can call it session, but then the name is ambiguous when used down in the function body: "is it an HTTP session? A database session? An Amazon session?".

Definitely naming it `db_session` would remove this ambiguity, and I don't see it as heavyweight in any way. Naming it `db` definitely makes me think it's a database instance and not a session (i.e. it makes me want to call `session` on `db`.)